sound/soc/sof/ipc4-loader.c

```diff
@@ -423,6 +423,65 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
 	return ret;
 }
 
+/**
+ * sof_ipc4_update_cpc_from_manifest - Update the cpc in base config from manifest
+ * @sdev: SOF device
+ * @fw_module: pointer struct sof_ipc4_fw_module to parse
+ * @basecfg: Pointer to the base_config to update
+ */
+void sof_ipc4_update_cpc_from_manifest(struct snd_sof_dev *sdev,
+				       struct sof_ipc4_fw_module *fw_module,
+				       struct sof_ipc4_base_module_cfg *basecfg)
+{
+	const struct sof_man4_module_config *fw_mod_cfg;
+	u32 cpc_pick = 0;
+	u32 max_cpc = 0;
+	int i;
+
+	if (!fw_module || !fw_module->fw_mod_cfg)
+		return;
+
+	/*
+	 * Find the best matching (highest) CPC value based on the module's
+	 * external configuration (IBS/OBS).
+	 * The CPC value in each module config entry has been measured and
+	 * recorded as a IBS/OBS/CPC triplet.
+	 *
+	 * In case of multiple IBS/OBS matches, use the highest CPC number.
+	 */
+	fw_mod_cfg = fw_module->fw_mod_cfg;
+	for (i = 0; i < fw_module->man4_module_entry.cfg_count; i++) {
+		if (basecfg->obs == fw_mod_cfg[i].obs &&
+		    basecfg->ibs == fw_mod_cfg[i].ibs &&
+		    cpc_pick < fw_mod_cfg[i].cpc)
+			cpc_pick = fw_mod_cfg[i].cpc;
+
+		if (max_cpc < fw_mod_cfg[i].cpc)
+			max_cpc = fw_mod_cfg[i].cpc;
+	}
+
+	basecfg->cpc = cpc_pick;
+
+	/*
+	 * No matching IBS/OBS found, use the maximum CPC value we have
+	 * encountered while parsing the module config entries
+	 */
+	if (!basecfg->cpc)
+		basecfg->cpc = max_cpc;
```

Comment:
I feel this is an error: the topology uses an ibs/obs value that is not aligned with the manifest. In that case, using the maximum cpc available is just inventing information. it's not reliable and it's not recommended.

I would just reject this configuration completely.

And see below why it's inconsistent ...

Reply:
We cannot reject this as nothing will work if a module has no matching mod_cfg entry, we print an info and that will give us the information what needs to be added to manifest.

Reply:
what makes you think the max is relevant or even works @ujfalusi ?

If for example we want to deal with 192kHz data, and somehow this leads to an IBS/OBS pair that is not listed in the manifest, it's possible that the max listed in the manifest is a low estimate of the actual needs.

So there's a strong likelihood that we would be trying our luck and fail later with random xruns.

My approach is always: fail big and fail early. Trying to make things work by accident is not a reliable solution.

Or at the very least add some linear fudge factor based on the ratio between IBS/OBS listed and IBS/OBS requested.

Reply:
OK, so treat the no exact match as no CPC value at all and use 0 (with a print)?